feat: add C implementation for `stats/base/dists/planck/stdev`

yuvimittal opened the pull request:
Resolves #4912 

## Description

> What is the purpose of this pull request?

This pull request:

-   Add C implementation for @stdlib/stats/base/dists/planck/stdev

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #4912  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: tweak permissions and use fine-grained PAT

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

yuvimittal committed:
Add C implementation for @stdlib/stats/base/dists/planck/stdev

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/README.md:
```diff
@@ -120,6 +120,95 @@ for ( i = 0; i < lambda.length; i++ ) {
+
+#### stdlib_base_dists_planck_stdev( lambda )
+
+Evaluates the standard deviation for an planck distribution.
```
```suggestion
Returns the standard deviation of a Planck distribution with shape parameter `λ`.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/benchmark/benchmark.js:
```diff
@@ -31,14 +32,19 @@ var stdev = require( './../lib' );
+	len = 100;
+	lambda = new Float64Array( len );
+	for ( i = 0; i < len; i++ ) {
+		lambda[ i ] = ( randu() * 10.0 ) + 1.0;
```
Will be needing `uniform` logic instead of `randu` as per C benchmark.
Same goes for the native version as well.

yuvimittal replied:
Can you explain why is that , also how to use that because as much i have seen randu is used .

Neerajpathak07 replied:
yeah, so the way you used `uniform` logic to generate values directly you need to do the same for the js file as well
and by using `uniform` you can generate values for specific range.
Apparently this was also mentioned by @kgryte in the Gitter channel.

yuvimittal replied:
Oh , alright !

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,138 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		lambda[i] = random_uniform( 0.1, 10.0 );
```
```suggestion
		lambda[i] = random_uniform( 1.0, 10.0 );
```

yuvimittal replied:
I think both 1 and 0.1 works , condition is lambda >0 , right ?

Neerajpathak07 replied:
yea, but just for added clarity!!

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/include/stdlib/stats/base/dists/planck/stdev.h:
```diff
@@ -0,0 +1,38 @@
+#endif
+
+/**
+* Evaluates the standard deviation (stdev) for an planck distribution with shape parameter lambda.
```
```suggestion
* Returns the standard deviation of a Planck distribution with shape parameter `λ`.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/manifest.json:
```diff
@@ -0,0 +1,91 @@
+{
+	"options": {
```
Hey, the `JSON` file should be intended with two spaces as per eslint rules.
I would usually end up doing the same mistake in the past so thought of letting you know.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/src/main.c:
```diff
@@ -0,0 +1,41 @@
+#include "stdlib/math/base/special/sqrt.h"
+
+/**
+* Evaluates the standard deviation of a Planck distribution given a shape parameter `lambda`.
```
```suggestion
*  Returns the standard deviation of a Planck distribution with shape parameter `λ`.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/planck/stdev/src/main.c:
```diff
@@ -0,0 +1,41 @@
+	if ( stdlib_base_is_nan( lambda ) || lambda <= 0.0 ) {
+		return 0.0/0.0; // NaN
+	}
+	double exp_lambda = stdlib_base_exp( -lambda ) ;
+	return stdlib_base_sqrt( exp_lambda ) / stdlib_base_abs( stdlib_base_expm1( -lambda ) );
```
```suggestion
	return stdlib_base_sqrt( stdlib_base_exp( -lambda )  ) / stdlib_base_abs( stdlib_base_expm1( -lambda ) );
```

Neerajpathak07 replied:
I think we should avoid creating a new variable as it increases the space complexity whereas we can avoid it. So, we can just wrap the C function for `exp` inside of `sqrt` itself.

yuvimittal replied:
the code is clearer this way , no ?

Neerajpathak07 commented:
Also change the title of your PR to:-
feat: add C implementation for `stats/base/dists/planck/stdev`
as it will throw PR lint error while CI testing
and add backticks to the starting and ending while naming packages!!

yuvimittal committed:
Readme update

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

yuvimittal committed:
update boundary condition

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
chore: update copyright year to 2025

Planeshifter reviewed (approved):
Thanks @yuvi-mittal for the PR and @Neerajpathak07 for review!
